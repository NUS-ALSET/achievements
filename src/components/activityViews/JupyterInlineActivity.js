/**
 * @file JupyterColabActivity container module
 * @author Theodor Shaytanov <theodor.shaytanov@gmail.com>
 * @created 08.03.18
 */

import React, { Fragment } from "react";
import PropTypes from "prop-types";

import cloneDeep from "lodash/cloneDeep";

import {
  problemSolutionRefreshFail,
  problemSolveUpdate
} from "../../containers/Activity/actions";
import { notificationShow } from "../../containers/Root/actions";
import JupyterNotebook from "./JupyterNotebook";
import Typography from "@material-ui/core/Typography";

class JupyterInlineActivity extends React.PureComponent {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    onChange: PropTypes.func,
    problem: PropTypes.object,
    solution: PropTypes.object,
    readOnly : PropTypes.bool
  };

  state = {
    solutionJSON: false
  };

  onSolutionRefreshClick = value => {
    const { dispatch, onChange, problem } = this.props;

    const solutionJSON = cloneDeep(problem.problemJSON);

    solutionJSON.cells[Number(problem.code)].source = value
      .split("\n")
      .map(line => line + "\n");

    this.setState({
      solutionJSON: solutionJSON || false
    });
    dispatch(problemSolutionRefreshFail());
    if (onChange) {
      onChange(solutionJSON);
    }
    if (!solutionJSON) {
      return dispatch(notificationShow("Code wasn't changed"));
    }

    return dispatch(
      problemSolveUpdate(problem.pathId, problem.problemId, solutionJSON)
    );
  };

  getCalculatedSolution = solution => {
    if (!solution) {
      return "";
    }
    if (solution.failed) {
      return (
        <Typography color="error">
          (Error- Final output should be empty)
        </Typography>
      );
    }
    if (solution.loading) {
      return <Typography color="textSecondary">(Checking)</Typography>;
    }
    if (solution.checked) {
      return <Typography color="primary">(Passed)</Typography>;
    }
  };

  // Move it to paths
  getSolutionCode = (solution, problem) =>
    (this.state.solutionJSON &&
      this.state.solutionJSON.cells &&
      this.state.solutionJSON.cells[Number(problem.code)].source
        .join("")
        .replace(/\n\n/g, "\n")) ||
    (solution &&
      solution.cells &&
      solution.cells[Number(problem.code)].source
        .join("")
        .replace(/\n\n/g, "\n")) ||
    (problem &&
      problem.problemJSON &&
      problem.problemJSON.cells &&
      problem.problemJSON.cells[Number(problem.code)].source
        .join("")
        .replace(/\n\n/g, "\n"));

  render() {
    const {
      /** @type {JupyterPathProblem} */
      problem,
      solution,
      readOnly
    } = this.props;
    
    return (
      <Fragment>
        <JupyterNotebook
          action={this.onSolutionRefreshClick}
          defaultValue={this.getSolutionCode(solution, problem)}
          persistent={true}
          richEditor={true}
          solution={false}
          readOnly={readOnly}
          title={readOnly ? "Submitted Code" : "Edit code"}
        />
        {solution &&
          (solution.json || solution.loading) && (
            <JupyterNotebook
              solution={solution}
              readOnly={readOnly}
              title={
                <Fragment>
                  Calculated Solution
                  {this.getCalculatedSolution(solution)}
                </Fragment>
              }
            />
          )}
        {solution &&
          solution.provided && (
            <JupyterNotebook
              readOnly={readOnly}
              solution={{ json: solution.provided }}
              title="Provided solution"
            />
          )}
        <JupyterNotebook
          readOnly={readOnly}
          solution={{ json: problem.problemJSON }}
          title="Problem"
        />
      </Fragment>
    );
  }
}

export default JupyterInlineActivity;
