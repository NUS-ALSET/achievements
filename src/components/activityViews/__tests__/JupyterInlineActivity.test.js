import React from "react";
import { createShallow } from "@material-ui/core/test-utils";

import sinon from "sinon";

import JupyterInlineActivity from "../JupyterInlineActivity";

describe("<JupyterInlineActivity>", () => {
  let mockDispatch;
  let shallow;

  beforeEach(() => {
    mockDispatch = sinon.spy();
    shallow = createShallow();
  });

  it("should check snapshot", () => {
    const wrapper = shallow(
      <JupyterInlineActivity
        dispatch={mockDispatch}
        problem={{
          problemJSON: "test"
        }}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("is just mock test", () => {
    const wrapper = shallow(
      <JupyterInlineActivity
        dispatch={mockDispatch}
        problem={{
          code: 0,
          problemJSON: {
            cells: [
              {
                source: []
              },
              "foobar"
            ]
          }
        }}
      />
    );

    wrapper.instance().onSolutionRefreshClick("test");
    // wrapper.instance().getSolutionCode({}, {});
    expect(wrapper.state("solutionJSON")).toEqual({
      cells: [{ source: ["test\n"] }, "foobar"]
    });
    expect(mockDispatch.called).toEqual(true);
  });
});
