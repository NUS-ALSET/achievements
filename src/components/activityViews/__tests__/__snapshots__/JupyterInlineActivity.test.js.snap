// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JupyterInlineActivity> should check snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <JupyterInlineActivity
    dispatch={[Function]}
    problem={
        Object {
            "problemJSON": "test",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <JupyterNotebook
          action={[Function]}
          defaultValue={undefined}
          persistent={true}
          readOnly={undefined}
          richEditor={true}
          solution={false}
          title="Edit code"
/>,
        undefined,
        undefined,
        <JupyterNotebook
          readOnly={undefined}
          solution={
                    Object {
                              "json": "test",
                            }
          }
          title="Problem"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "action": [Function],
          "defaultValue": undefined,
          "persistent": true,
          "readOnly": undefined,
          "richEditor": true,
          "solution": false,
          "title": "Edit code",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      undefined,
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "readOnly": undefined,
          "solution": Object {
            "json": "test",
          },
          "title": "Problem",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <JupyterNotebook
            action={[Function]}
            defaultValue={undefined}
            persistent={true}
            readOnly={undefined}
            richEditor={true}
            solution={false}
            title="Edit code"
/>,
          undefined,
          undefined,
          <JupyterNotebook
            readOnly={undefined}
            solution={
                        Object {
                                    "json": "test",
                                  }
            }
            title="Problem"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "action": [Function],
            "defaultValue": undefined,
            "persistent": true,
            "readOnly": undefined,
            "richEditor": true,
            "solution": false,
            "title": "Edit code",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        undefined,
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "readOnly": undefined,
            "solution": Object {
              "json": "test",
            },
            "title": "Problem",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {},
  },
}
`;
